tests/src/test/scala/com/nvidia/spark/rapids/GpuKudoWritePartitioningSuite.scala

```diff
@@ -0,0 +1,511 @@
+      Map.empty,
+      None,
+      Seq.empty,
+      enableOpTimeTrackingRdd = false)
+    dependency
+  }
+
+  /**
+   * Collects batches from the shuffle dependency, optionally injecting OOM for retry testing
+   * Returns: (batches, totalRowsSeen, numNextCalls)
+   */
+  private def collectBatches(
+      dependency: org.apache.spark.ShuffleDependency[Int, ColumnarBatch, ColumnarBatch],
+      injectOOM: Boolean): (mutable.ArrayBuffer[(Int, ColumnarBatch)], Long, Int) = {
+    val allPartitionedBatches = mutable.ArrayBuffer[(Int, ColumnarBatch)]()
+    var totalRowsSeen = 0L
+    var firstIteration = true
+    var numNextCalls = 0
+
+    if (injectOOM) {
+      // Associate the current thread with a task (required for OOM injection)
+      RmmSpark.currentThreadIsDedicatedToTask(1)
+    }
+
+    try {
+      dependency.rdd.partitions.foreach { partition =>
+        val partitionIterator = dependency.rdd.iterator(partition,
+          org.apache.spark.TaskContext.get())
+
+        // Inject a split OOM right before the first call to next() if requested
+        if (injectOOM && firstIteration && partitionIterator.hasNext) {
+          val threadId = RmmSpark.getCurrentThreadId
+          RmmSpark.forceSplitAndRetryOOM(threadId, 1,
+            RmmSpark.OomInjectionType.GPU.ordinal, 0)
+          firstIteration = false
+        }
+
+        while (partitionIterator.hasNext) {
+          val result = partitionIterator.next()
+          numNextCalls += 1
+          val partitionId = result._1
+          val batch = result._2
+
+          // Increment ref counts before collecting (batches are managed by rddWithPartitionIds
+          // and will be closed automatically, so we need to take ownership)
+          if (batch.numCols() > 0 &&
+              batch.column(0).isInstanceOf[SlicedSerializedColumnVector]) {
+            SlicedSerializedColumnVector.incRefCount(batch)
+          } else {
+            GpuColumnVector.incRefCounts(batch)
+          }
+          allPartitionedBatches.append((partitionId, batch))
+          totalRowsSeen += batch.numRows()
+        }
+      }
+    } finally {
+      if (injectOOM) {
+        RmmSpark.removeAllCurrentThreadAssociation()
+      }
+    }
+
+    (allPartitionedBatches, totalRowsSeen, numNextCalls)
+  }
+
+  /**
+   * Verifies that the collected batches match the expected test data.
+   * Note: This method does NOT close the input batches - caller is responsible for cleanup.
+   */
+  private def verifyBatchContents(
+      allPartitionedBatches: mutable.ArrayBuffer[(Int, ColumnarBatch)],
+      totalRowsSeen: Long,
+      serializer: GpuColumnarBatchSerializer): Unit = {
+    // Verify total rows are preserved
+    assert(totalRowsSeen == expectedTotalRows,
+      s"Expected $expectedTotalRows total rows (10 per batch × 2 batches), got $totalRowsSeen")
+
+    // Deserialize SlicedSerializedColumnVector batches
+    // Note: We create a copy of the batch references to avoid closing the originals
```

Comment:
I am confused by this comment. Where do we copy the batches? I see we create a sequence, but what does that have to do with closing?